text/0000-named-custom-cargo-profiles.md

```diff
@@ -0,0 +1,275 @@
+* A `debug` profile name is not allowed, with a warning saying to use the
+  already established `dev` name.
+* The `bench` profile defaults to the following definition, to preserve current
+  behavior:
+
+```
+[profile.bench]
+inherits = "release"
+dir-name = "release"
+```
+
+* The `test` profile defaults to the following definition, to preserve current behavior:
+```
+[profile.test]
+inherits = "dev"
+dir-name = "debug"
+```
+
+* The `doc` profile defaults to the following definition, to preserve current
+  behavior:
+```
+[profile.doc]
+inherits = "dev"
+dir-name = "debug"
+```
+
+
+* The (upcoming) `build` profile defaults to the following definition:
+```
+[profile.build]
+inherits = "dev"
+dir-name = "debug"
+debug = false
+```
+
+
+## Treatment to Cargo's 'Finished' print
+
+For `cargo build`, the profile's name is emitted in the "Finished" line:
+
+    Finished release-lto [optimized + lto] target(s) in 3.83s
+
+* As before, optimization mode is always printed, with `optimized` if
+  `opt-level > 0` and `unoptimized` othersize.
+* As before, information regarding debuginfo is printed, with `debuginfo` if
+  `debug != false`, and nothing if `debug == false` (or `debug == 0`).
+* Other differences from the inherited root profile pre-defined defaults are
+  printed in a concise manner.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The 'final outputs' phrasing was used in this RFC, knowing that there are
+intermediate outputs that live under `target/` that are usually not a concern
+for most Cargo users. The paths that constitue the final build outputs however,
+constitute as sort of a protocol for invokers of Cargo. This RFC extends on
+that protocol, allowing for outputs in more directories.
+
+## Cargo code changes
+
+In implementation details, there are various hand-coded references to pre-defined
+profiles, that we would like to remove.
+
+The `BuildConfig` structure currently has a `release` boolean. The
+implementation will replace it with a value of type `enum Profile {Dev,
+Release, Custom(String))`.
+
+* The `Profiles` struct in `cargo/core/profiles.rs` currently has hardcoded
+  `dev`, `release`, `test`, `bench`, and `doc` fields. This should be changed
+  into a `BTreeMap` based on profile names. The pre-defined profiles can be
+  loaded into it, before `TomlProfile` overrides are applied to them.
+* Similarly, `TomlProfiles` will be changed to hold profiles in a `BTreeMap`.
+* We would need to compute the actual `build_override` for a profile based on
+  resolution through the `inherited` key.
+* Custom build scripts: For compatibility, the `PROFILE` environment currently
+  being passed to the `build.rs` script is going to bear the name of the output
+  directory and not the name of the profile.
+
+## Collision under the target directory
+
+To guard against the unlikely case that a target name is identical to a profile
```

Comment:
I'm not sure this is necessary, since target names have a particular format that is unlikely to clash, and it would require collecting that information (it seems to add a nontrivial performance cost).  It seems unlikely someone would try to force a collision, and the consequences are not dire.  The most likely situation where there is a collision is with a custom target definition, so maybe checking that would be reasonable (if someone specifies `--target foo.json`).

There should be an exclusion list. The entries I can think of now is: "package", "build", "debug", "doc", anything starting with ".".

Reply:
Agreed. It would also be easier to implement. I'll add the exclusion list suggestion to the RFC.